File: server/services/villageService.js
const fs = require("fs");
const path = require("path");
const axios = require("axios");

const VillageModel = require("../models/Village.js");

const searchCache = new Map();
const boundaryCache = new Map();
const searchIndex = [];

function unwrapGeometry(value) {
  if (!value || typeof value !== "object") return null;
  if (value.type === "Feature") return value.geometry || null;
  if (value.type === "FeatureCollection") {
    return value.features?.[0]?.geometry || null;
  }
  return value;
}

function getOuterRing(geometry) {
  const normalized = unwrapGeometry(geometry);
  if (!normalized || !Array.isArray(normalized.coordinates)) return null;

  const coordinates = normalized.coordinates;
  if (normalized.type === "Polygon") {
    return Array.isArray(coordinates[0]) ? coordinates[0] : null;
  }
  if (normalized.type === "MultiPolygon") {
    return Array.isArray(coordinates[0]?.[0]) ? coordinates[0][0] : null;
  }
  return null;
}

function computePolygonAreaSqm(geometry) {
  // Basic bounding box approximation for area in square meters.
  const coords = getOuterRing(geometry);
  if (!coords || coords.length === 0) return 1000000;

  let minX = Infinity,
    maxX = -Infinity,
    minY = Infinity,
    maxY = -Infinity;
  for (const p of coords) {
    if (!Array.isArray(p) || p.length < 2) continue;
    if (p[0] < minX) minX = p[0];
    if (p[0] > maxX) maxX = p[0];
    if (p[1] < minY) minY = p[1];
    if (p[1] > maxY) maxY = p[1];
  }
  if (!Number.isFinite(minX) || !Number.isFinite(minY)) return 1000000;
  const widthM =
    (maxX - minX) * 111000 * Math.cos(((minY + maxY) / 2) * (Math.PI / 180));
  const heightM = (maxY - minY) * 111000;
  return Math.max(100000, Math.abs(widthM * heightM * 0.75));
}

async function loadVillages() {
  try {
    searchCache.clear();
    boundaryCache.clear();
    searchIndex.length = 0;

    const dataDir = path.resolve("data");
    const geojsonPath = path.join(dataDir, "maharashtra.geojson");

    if (fs.existsSync(geojsonPath)) {
      const raw = fs.readFileSync(geojsonPath, "utf-8");
      const parsed = JSON.parse(raw);
      const features = parsed.features || [];

      let count = 0;
      for (const feat of features) {
        const props = feat.properties || {};
        const geom = feat.geometry || {};
        if (!geom || !["Polygon", "MultiPolygon"].includes(geom.type)) continue;

        const v_id = String(props.id || count + 1);
        const name = props.name || "Unknown";
        const district = props.district || "Unknown";
        const state = props.state || "Maharashtra";

        let lat = 20.0,
          lon = 78.0;
        const coords = geom.coordinates;
        if (coords && coords.length > 0) {
          const pts =
            geom.type === "MultiPolygon" ? coords[0]?.[0] || [] : coords[0];
          if (pts.length > 0) {
            lon = pts.reduce((sum, p) => sum + p[0], 0) / pts.length;
            lat = pts.reduce((sum, p) => sum + p[1], 0) / pts.length;
          }
        }

        const areaHa = computePolygonAreaSqm(geom) / 10000.0;

        const village = {
          id: v_id,
          name,
          nameHindi: name,
          district,
          state,
          coordinates: [lat, lon],
          lat,
          lon,
          boundary: geom,
          area: areaHa,
          soilType: props.soilType || "Black Cotton / Clay Loam",
          primaryCrops: ["Soybean", "Cotton", "Sugarcane", "Chickpea"],
        };

        searchCache.set(v_id, village);
        boundaryCache.set(v_id, geom);
        searchIndex.push({
          id: v_id,
          name,
          district,
          state,
        });

        // Async seed into MongoDB
        VillageModel.updateOne(
          { id: v_id },
          { $set: village },
          { upsert: true },
        ).catch(() => {});

        count++;
      }
      console.log(
        `[VillageService] Loaded ${count} villages into index & DB cache.`,
      );
    }
  } catch (err) {
    console.error("[VillageService Error]", err.message);
  }
}

async function searchVillages(query) {
  if (!query) return searchIndex.slice(0, 20);
  const q = query.toLowerCase();

  // Primary local index search
  const matches = searchIndex.filter((item) =>
    item.name.toLowerCase().includes(q),
  );
  if (matches.length > 0) {
    return matches.slice(0, 20);
  }

  // Fallback OSM Nominatim Geocoding
  try {
    const resp = await axios.get("https://nominatim.openstreetmap.org/search", {
      params: {
        q: `${query}, India`,
        format: "json",
        polygon_geojson: 1,
        limit: 5,
      },
      headers: { "User-Agent": "GramDrishti-Farmer-Platform/1.0" },
      timeout: 5000,
    });

    if (resp.data && resp.data.length > 0) {
      const results = [];
      for (const item of resp.data) {
        const v_id = `osm_${item.place_id}`;
        const name = item.display_name.split(",")[0] || query;
        const district =
          item.display_name.split(",")[1]?.trim() || "Custom District";
        const polygon = item.geojson || {
          type: "Polygon",
          coordinates: [
            [
              [parseFloat(item.lon) - 0.02, parseFloat(item.lat) - 0.02],
              [parseFloat(item.lon) + 0.02, parseFloat(item.lat) - 0.02],
              [parseFloat(item.lon) + 0.02, parseFloat(item.lat) + 0.02],
              [parseFloat(item.lon) - 0.02, parseFloat(item.lat) + 0.02],
              [parseFloat(item.lon) - 0.02, parseFloat(item.lat) - 0.02],
            ],
          ],
        };
        addDynamicVillage(v_id, polygon, name, district);
        results.push({ id: v_id, name, district, state: "India" });
      }
      return results;
    }
  } catch (e) {
    console.warn("[Nominatim Search Warning]", e.message);
  }

  return searchIndex.slice(0, 10);
}

function getVillageById(id) {
  const v_id = String(id);
  return searchCache.get(v_id) || searchCache.get("1") || null;
}

function getVillageBoundary(id) {
  const v_id = String(id);
  return boundaryCache.get(v_id) || boundaryCache.get("1") || null;
}

async function addDynamicVillage(
  v_id,
  polygon,
  name = "Custom Farm Location",
  district = "Dynamic",
) {
  const geometry = unwrapGeometry(polygon) || polygon;
  let lat = 20.0,
    lon = 78.0;
  const coords = getOuterRing(geometry) || [];
  if (coords.length > 0) {
    const validPoints = coords.filter(
      (point) => Array.isArray(point) && point.length >= 2,
    );
    if (validPoints.length > 0) {
      lon = validPoints.reduce((sum, p) => sum + p[0], 0) / validPoints.length;
      lat = validPoints.reduce((sum, p) => sum + p[1], 0) / validPoints.length;
    }
  }

  const areaHa = computePolygonAreaSqm(geometry) / 10000.0;
  const village = {
    id: v_id,
    name: name || v_id,
    nameHindi: name || v_id,
    district,
    state: "Maharashtra",
    coordinates: [lat, lon],
    lat,
    lon,
    boundary: geometry,
    area: areaHa,
    soilType: "Medium Black Soil",
    primaryCrops: ["Soybean", "Wheat", "Gram", "Cotton"],
  };

  searchCache.set(v_id, village);
  boundaryCache.set(v_id, geometry);

  VillageModel.updateOne(
    { id: v_id },
    { $set: village },
    { upsert: true },
  ).catch(() => {});
  return village;
}

function getSearchIndex() {
  return searchIndex;
}

function getAllVillageBoundaries() {
  const features = [];
  for (const [id, village] of searchCache.entries()) {
    if (village.boundary) {
      features.push({
        type: "Feature",
        id,
        properties: {
          id,
          name: village.name,
          district: village.district,
          state: village.state,
        },
        geometry: village.boundary,
      });
    }
  }
  return {
    type: "FeatureCollection",
    features,
  };
}

module.exports = {
  loadVillages,
  searchVillages,
  getVillageById,
  getVillageBoundary,
  addDynamicVillage,
  getSearchIndex,
  getAllVillageBoundaries,
};

File: server/app/services/village_service.py
import json
import os
import glob
from typing import List, Dict, Any, Optional
from app.models.village import Village
from app.core.logging import get_logger
from app.services.gee.geometry import compute_polygon_area_sqm

logger = get_logger(__name__)

# Cache for villages: { "id": Village }
SEARCH_CACHE: Dict[str, Village] = {}
# Cache for village boundaries: { "id": Dict[str, Any] }
BOUNDARY_CACHE: Dict[str, Dict[str, Any]] = {}
# Search index: List of searchable dictionaries without full boundary
SEARCH_INDEX: List[Dict[str, Any]] = []


def load_villages():
    """Load all GeoJSON files from data/ directory and build the index."""

    SEARCH_CACHE.clear()
    BOUNDARY_CACHE.clear()
    SEARCH_INDEX.clear()

    data_dir = os.path.join(
        os.path.dirname(
            os.path.dirname(
                os.path.dirname(__file__))),
        "data")
    geojson_files = glob.glob(os.path.join(data_dir, "*.geojson"))

    logger.info(f"Loading GeoJSON files from {data_dir}...")

    count = 0
    for file_path in geojson_files:
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)
                features = data.get("features", [])

                for feature in features:
                    props = feature.get("properties", {})
                    geom = feature.get("geometry", {})

                    if not geom or geom.get("type") not in [
                            "Polygon", "MultiPolygon"]:
                        continue

                    v_id = str(props.get("id", count))
                    name = props.get("name", "Unknown")
                    district = props.get("district", "Unknown")
                    state = props.get("state", "Unknown")

                    # Calculate approximate center from coordinates for display
                    # purposes
                    coords = geom.get("coordinates", [])
                    poly_type = geom.get("type", "Polygon")
                    lat = 20.0
                    lon = 78.0
                    if coords and len(coords) > 0:
                        if poly_type == "MultiPolygon":
                            if len(coords[0]) > 0 and len(coords[0][0]) > 0:
                                pts = coords[0][0]
                                lon = sum([p[0] for p in pts]) / len(pts)
                                lat = sum([p[1] for p in pts]) / len(pts)
                        elif poly_type == "Polygon":
                            if len(coords[0]) > 0:
                                pts = coords[0]
                                lon = sum([p[0] for p in pts]) / len(pts)
                                lat = sum([p[1] for p in pts]) / len(pts)

                    real_area_ha = compute_polygon_area_sqm(geom) / 10000.0 if geom else 0.0

                    village = Village(
                        id=v_id,
                        name=name,
                        nameHindi=name,
                        district=district,
                        state=state,
                        coordinates=(lat, lon),
                        boundary=geom,
                        area=real_area_ha
                    )

                    SEARCH_CACHE[v_id] = village
                    BOUNDARY_CACHE[v_id] = geom

                    # Add to search index (without heavy geometry)
                    SEARCH_INDEX.append({
                        "id": v_id,
                        "name": name,
                        "district": district,
                        "state": state
                    })

                    count += 1
        except Exception as e:
            logger.error(f"Failed to load {file_path}: {e}")

    logger.info(f"Loaded {count} villages into the index.")


async def search_villages(query: str) -> List[Dict[str, Any]]:
    """Search for villages in the loaded index by name."""
    query_lower = query.lower()
    results = []

    for item in SEARCH_INDEX:
        # Match primarily on name so general state/district matches don't suppress Nominatim
        if query_lower in item["name"].lower():
            results.append(item)
            if len(results) >= 20:  # Limit results
                break

    return results


def get_village_by_id(village_id: str) -> Optional[Village]:
    """Return a specific village by ID."""
    return SEARCH_CACHE.get(village_id)


def get_village_boundary(village_id: str) -> Optional[Dict[str, Any]]:
    """Return the GeoJSON boundary for a specific village."""
    return BOUNDARY_CACHE.get(village_id)


def add_dynamic_village(
        village_id: str, polygon: Dict[str, Any], name: str = None) -> Village:
    """Add a dynamically fetched polygon (e.g. from Nominatim) to the cache."""
    # Calculate approximate center
    lat = 20.0
    lon = 78.0
    coords = polygon.get("coordinates", [])
    poly_type = polygon.get("type", "Polygon")

    if coords and len(coords) > 0:
        if poly_type == "MultiPolygon":
            # For MultiPolygon, coords[0] is a Polygon (list of rings)
            # coords[0][0] is the exterior ring (list of points)
            if len(coords[0]) > 0 and len(coords[0][0]) > 0:
                pts = coords[0][0]
                lon = sum([p[0] for p in pts]) / len(pts)
                lat = sum([p[1] for p in pts]) / len(pts)
        elif poly_type == "Polygon":
            # For Polygon, coords[0] is the exterior ring (list of points)
            if len(coords[0]) > 0:
                pts = coords[0]
                lon = sum([p[0] for p in pts]) / len(pts)
                lat = sum([p[1] for p in pts]) / len(pts)

    real_area_ha = compute_polygon_area_sqm(polygon) / 10000.0 if polygon else 0.0

    village = Village(
        id=village_id,
        name=name or village_id,
        nameHindi=name or village_id,
        district="Dynamic",
        state="Dynamic",
        coordinates=(lat, lon),
        boundary=polygon,
        area=real_area_ha
    )

    SEARCH_CACHE[village_id] = village
    BOUNDARY_CACHE[village_id] = polygon
    return village
